cdc/processor/sinkmanager/table_sink_worker.go

```diff
@@ -81,48 +82,218 @@ func (w *sinkWorker) handleTasks(ctx context.Context, taskChan <-chan *sinkTask)
+		// we can wait for next round.
+		if txnFinished {
+			if w.sinkMemQuota.TryAcquire(requestMemSize) {
+				*availableMem += requestMemSize
+				log.Debug("MemoryQuotaTracing: try acquire memory for table sink task",
+					zap.String("namespace", w.changefeedID.Namespace),
+					zap.String("changefeed", w.changefeedID.ID),
+					zap.Stringer("span", &task.span),
+					zap.Uint64("memory", requestMemSize))
+			}
+		} else {
+			// The transaction is not finished and splitTxn is false, we need to
+			// force acquire memory. Because we can't leave rest data
+			// to the next round.
+			if !w.splitTxn {
+				w.sinkMemQuota.ForceAcquire(requestMemSize)
+				*availableMem += requestMemSize
+				log.Debug("MemoryQuotaTracing: force acquire memory for table sink task",
+					zap.String("namespace", w.changefeedID.Namespace),
+					zap.String("changefeed", w.changefeedID.ID),
+					zap.Stringer("span", &task.span),
+					zap.Uint64("memory", requestMemSize))
+			} else {
+				// NOTE: if splitTxn is true it's not required to force acquire memory.
+				// We can wait for a while because we already flushed some data to
+				// the table sink.
+				if err := w.sinkMemQuota.BlockAcquire(requestMemSize); err != nil {
+					return errors.Trace(err)
+				}
+				*availableMem += requestMemSize
+				log.Debug("MemoryQuotaTracing: block acquire memory for table sink task",
+					zap.String("namespace", w.changefeedID.Namespace),
+					zap.String("changefeed", w.changefeedID.ID),
+					zap.Stringer("span", &task.span),
+					zap.Uint64("memory", requestMemSize))
+			}
+		}
+	}
+	return nil
+}
+
+func (w *sinkWorker) handleTask(ctx context.Context, task *sinkTask) (finalErr error) {
+	lowerBound, upperBound := validateAndAdjustBound(w.changefeedID, task)
+
 	// First time to run the task, we have initialized memory quota for the table.
 	availableMem := requestMemSize
 	usedMem := uint64(0)
 
-	// Used to record the last written position.
-	// We need to use it to update the lower bound of the table sink.
-	var lastPos engine.Position
-
-	// To advance table sink in different cases: splitTxn is true or not.
-	committedTxnSize := uint64(0) // Can be used in `advanceTableSink`.
-	pendingTxnSize := uint64(0)   // Can be used in `advanceTableSinkWithBatchID`.
-	lastTxnCommitTs := uint64(0)  // Can be used in `advanceTableSink`
-	currTxnCommitTs := uint64(0)  // Can be used in `advanceTableSinkWithBatchID`.
-	events := make([]*model.RowChangedEvent, 0, 1024)
-
 	// batchID is used to advance table sink with a given CommitTs, even if not all
 	// transactions with the same CommitTs are collected, regardless of whether splitTxn
 	// is enabled or not. We split transactions with the same CommitTs even if splitTxn
-	// is false, and it won't break transaction atomicity to downstreams.
+	// is false, and it won't break transaction atomicity to downstream.
 	batchID := uint64(1)
-
+	// ？？？ What does drained mean?
```

Comment:
Seems it means we know we can't fetch more events from engine, because all available events are fetched from cache. So we can return directly, to skip creating an engine iterator.

Reply:
Oh, I forget to delete this comment! Thank you!